File: nodejs-js/chapter-5/github/authcode.js
/**
 * @file authcode.js
 * @description
 * This file implements GitHub OAuth 2.0 Authorization Code flow using Express.js.
 * It provides endpoints for login, logout, callback handling, and fetching user resources.
 * The server runs over HTTPS and uses cookies to store access tokens securely.
 *
 * @endpoint
 * GET /hello
 *   - Returns a simple "Hello, World!" message.
 *
 * GET /oauth/login
 *   - Initiates the OAuth 2.0 login flow with GitHub.
 *   - Redirects the user to GitHub's authorization endpoint.
 *
 * GET /oauth/logout
 *   - Logs out the user by clearing the authentication token cookie.
 *   - Redirects to the home page.
 *
 * GET /oauth/callback
 *   - Handles the OAuth 2.0 callback from GitHub.
 *   - Exchanges the authorization code for an access token.
 *   - Stores the access token in a secure, HTTP-only cookie.
 *   - Redirects to the home page.
 *
 * GET /resource
 *   - Fetches the authenticated user's profile information from GitHub.
 *   - Requires a valid access token in the cookie.
 *   - Returns user information as JSON.
 *
 * @note
 * - Uses HTTPS with custom certificates.
 * - Requires environment variables GH_CLIENT_ID and GH_CLIENT_SECRET for GitHub OAuth.
 * - Uses the openid-client library for OAuth operations.
 */
const express = require("express");
const https = require("node:https");
const app = express();
const cookieParser = require('cookie-parser')
app.use(cookieParser())
app.use(express.static('frontend'));
app.get('/hello', (req, res) => {
  res.send('Hello, World!');
});

{
  const oclient = require("openid-client");

  const cid = process.env.GH_CLIENT_ID;
  const csecret = process.env.GH_CLIENT_SECRET;

  const config = new oclient.Configuration({
    issuer: "https://github.com",
    authorization_endpoint: "https://github.com/login/oauth/authorize",
    token_endpoint: "https://github.com/login/oauth/access_token"
  }, cid, csecret); 

  const states = new Set();

  app.get('/oauth/login', (req, res) => {
    const state = oclient.randomState();
    const params = {
      state,
      redirect_uri: "https://mysrv.local:8443/oauth/callback",
      scope: ["user"]
    }
    const url = oclient.buildAuthorizationUrl(config, params);
    states.add(state);
    res.redirect(url.href);
  });

  app.get('/oauth/logout', (req, res) => {
    res.clearCookie('token');
    res.redirect('/');
  });

  app.get('/oauth/callback', (req, res) => {
    const state = req.query.state;
    if (state && states.has(state)) {
      states.delete(state);
      if (req.query.error) {
        res.status(401).send(req.query.error_description);
        return;
      }

      surl = req.protocol+"://"+req.host+req.originalUrl;
      console.log(surl);
      url = new URL(surl);

      const params = {
        redirect_uri: "https://mysrv.local:8443/oauth/callback",
        scope: ["user"]
      }
      oclient.authorizationCodeGrant(config, url, {
        expectedState: state
      }, params).then((tokens) => {
        console.log(tokens)
        res.cookie('token', tokens.access_token, {
          httpOnly: true,
          secure: true,
          path: '/'
        });
        res.redirect('/');
      }).catch((e) => {
        console.log(e);
        res.status(401).send(e.message);
      })
    } else {
      res.status(400).send("The server received a bad request");
    }
  });

  app.get('/resource', (req, res) => {
    console.log(req.cookies)
    token = req.cookies.token
    if (!token){
      res.status(401).send("User not authenticated");
    } else {
      console.log(token);
      oclient.fetchProtectedResource(config, token, new URL('https://api.github.com/user'), 'GET')
        .then(response => {
          if (!response.ok) 
            res.status(500).send("Unable to collect user information")
          else 
            return response.json();
        })
        .then(v => {
          console.log(v);
          res.json(v);
        })
        .catch(e => console.error('Error:', e));
    }
  });
}

const path = require("path");
const scertpath = path.join("..", "certs");
const fs = require("fs");

var tlsopts = {
  key:  fs.readFileSync(path.join(scertpath, "mysrv.local.key")),
  passphrase: "password",
  cert: Buffer.concat([
    fs.readFileSync(path.join(scertpath, "mysrv.local.crt")),
    fs.readFileSync(path.join(scertpath, "sint.crt"))
  ])
};

const httpsServer = https.createServer(tlsopts, app);

httpsServer.listen(8443, () => {
  console.log("HTTPS server up and running on port 8443")
});

File: python-js/chapter-5/github/authcode.py
# Flask app implementing GitHub OAuth2 Authorization Code flow for dev use.
# Requires GH_CLIENT_ID and GH_CLIENT_SECRET environment variables.
# Serves static frontend files and a simple index at '/'.
# /oauth/login redirects to GitHub authorize endpoint (scope: read:user).
# /oauth/callback exchanges the code for a JSON access token and stores
# it in an HttpOnly, Secure cookie named 'gh_access_token' (SameSite=Lax).
# /oauth/logout clears the access token cookie.
# /resource reads the cookie and calls GitHub's user API with the token.
# Builds a TLS chain by concatenating server cert and intermediate CA, then
# starts Flask on port 8443 with a TLS context using the chain file and key.
# Notes: short request timeouts, hard-coded redirect_uri and cert paths, and
# storing raw tokens in cookies is suitable only for demos, not production.

import os
import ssl
from flask import Flask, redirect, request, make_response, jsonify
import requests
from urllib.parse import urlencode

app = Flask(__name__)

# Configuration
GH_CLIENT_ID = os.environ.get('GH_CLIENT_ID')
GH_CLIENT_SECRET = os.environ.get('GH_CLIENT_SECRET')

if not GH_CLIENT_ID or not GH_CLIENT_SECRET:
  raise SystemExit("GH_CLIENT_ID and GH_CLIENT_SECRET environment variables must be set and non-blank.")

GH_AUTH_URL = "https://github.com/login/oauth/authorize"
GH_TOKEN_URL = "https://github.com/login/oauth/access_token"
GH_USER_API = "https://api.github.com/user"
REDIRECT_URI = "https://mysrv.local:8443/oauth/callback"
COOKIE_NAME = "gh_access_token"

# Endpoints

app.static_folder = os.path.abspath(os.path.join(os.path.dirname(__file__), 'frontend'))

@app.route('/hello')
def hello():
  return "Hello, World!"

@app.route('/')
def index():
  return app.send_static_file('index.html'), 200

@app.route('/<path:filename>')
def static_files(filename):
  return app.send_static_file(filename), 200

@app.route('/oauth/login')
def oauth_login():
  params = {
    "client_id": GH_CLIENT_ID,
    "redirect_uri": REDIRECT_URI,
    "scope": "read:user",
    "response_type": "code"
  }
  url = f"{GH_AUTH_URL}?{urlencode(params)}"
  return redirect(url)

@app.route('/oauth/logout')
def oauth_logout():
  resp = make_response(redirect('/'))
  resp.set_cookie(COOKIE_NAME, '', expires=0, httponly=True, secure=True, samesite='Lax')
  return resp

@app.route('/oauth/callback')
def oauth_callback():
  code = request.args.get('code')
  if not code:
    return "Missing code", 400

  data = {
    "client_id": GH_CLIENT_ID,
    "client_secret": GH_CLIENT_SECRET,
    "code": code,
    "redirect_uri": REDIRECT_URI
  }
  headers = {'Accept': 'application/json'}
  token_resp = requests.post(GH_TOKEN_URL, data=data, headers=headers, timeout=10)
  token_json = token_resp.json()
  access_token = token_json.get('access_token')
  if not access_token:
    return "Failed to obtain access token", 400

  resp = make_response(redirect('/'))
  resp.set_cookie(COOKIE_NAME, access_token, httponly=True, secure=True, samesite='Lax')
  return resp

@app.route('/resource')
def resource():
  access_token = request.cookies.get(COOKIE_NAME)
  if not access_token:
    return jsonify({"error": "Unauthorized"}), 401

  headers = {
    "Authorization": f"token {access_token}",
    "Accept": "application/json"
  }
  user_resp = requests.get(GH_USER_API, headers=headers, timeout=10)
  if user_resp.status_code != 200:
    return jsonify({"error": "Failed to fetch user info"}), 400
  return jsonify(user_resp.json())

# Update SSL certificate and key paths        
CERTS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'certs'))
SERVER_CERT = os.path.join(CERTS_DIR, 'mysrv.local.crt')
SERVER_KEY = os.path.join(CERTS_DIR, 'mysrv.local.key')
CA_CERT = os.path.join(CERTS_DIR, 'sint.crt')

# Combine server cert and intermediate CA into a chain file
CHAIN_CERT = os.path.join(CERTS_DIR, 'mysrv.local.chain.crt')
if not os.path.exists(CHAIN_CERT):
  with open(CHAIN_CERT, 'w') as chain:
    with open(SERVER_CERT, 'r') as sc:
      chain.write(sc.read())
    with open(CA_CERT, 'r') as ca:
      chain.write('\n')
      chain.write(ca.read())
      
if __name__ == '__main__':
  context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
  context.load_cert_chain(CHAIN_CERT, SERVER_KEY, password="password")
  app.run(host='0.0.0.0', port=8443, ssl_context=context)